### PR TITLE
Feat/versioned initializable

### PR DESCRIPTION
**Pull request by msaug**
Implementation of the VersionedInitializable contracts with minor changes.
- VersionedInitializable is used to make sure that the same implementation of a Proxy is never initialized twice.
- The re-entrancy possibility inside the `initializer` modifier in Solidity has been disabled (no longer possible to call `initialize` inside `initialize`
- The "mock" files are proxy implementation mocks used in the upgradeability tests
- The tests should be improved with the implementation of InitializableImmutableAdminUpgradeabilityProxy.
- I created another test file in the `with_deployment` folder because the contracts used in this test file are not used in other parts of the test suites.

### CODE REVIEW

File: contracts/protocol/libraries/aave_upgradeability/versioned_initializable.cairo [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        is_top_level_call : felt
+    ) -> ():
+        if is_top_level_call == TRUE:
+            VersionedInitializable_initializing.write(FALSE)
+            return ()
+        end
+        return ()
+    end
+
+    func get_revision{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}() -> (
```
Comment (hsouf) [inline contracts/protocol/libraries/aave_upgradeability/versioned_initializable.cairo]: we shouldn't have a get_revision function here. The current revision should be provided as input from the implementation contract.

File: contracts/protocol/libraries/aave_upgradeability/versioned_initializable.cairo [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            let (requirement_2) = BoolCompare.either(initializing, is_current_gt_last)
+            let (requirement_final) = BoolCompare.either(requirement_1, requirement_2)
+            assert requirement_final = TRUE
+        end
+
+        let (is_top_level_call) = BoolCompare.not(initializing)
+
+        if is_top_level_call == TRUE:
+            VersionedInitializable_initializing.write(TRUE)
+            VersionedInitializable_last_initialized_revision.write(current_revision)
```
Comment (hsouf) [inline contracts/protocol/libraries/aave_upgradeability/versioned_initializable.cairo]: VersionedInitializable_last_initialized_revision.write(1)

File: contracts/protocol/libraries/aave_upgradeability/versioned_initializable.cairo [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+end
+
+namespace VersionedInitializable:
+    func _before_initialize{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        ) -> (is_top_level_call : felt):
+        alloc_locals
+        let (last_revision) = VersionedInitializable_last_initialized_revision.read()
+        let (current_revision) = get_revision()
+        let (initializing) = VersionedInitializable_initializing.read()
+        # TODO find a way to detect its usage in a constructor
```
Comment (hsouf) [inline contracts/protocol/libraries/aave_upgradeability/versioned_initializable.cairo]: we should probably remove the `is constructor` check, we don't have the privilege to do it on Starknet (as far as I'm concerned).

**Comment by hsouf** [comment]
@msaug  sorry I didn't notice it was a draft PR, just noticed that some comments are being addressed at the time of writing ^^

File: tests/contracts/mock_aave_upgradeable_proxy.cairo [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+# OpenZeppelin Contracts for Cairo v0.2.0 (upgrades/Proxy.cairo)
+
+%lang starknet
+# %builtins pedersen range_check bitwise
+
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+from starkware.starknet.common.syscalls import library_call, library_call_l1_handler
+from openzeppelin.upgrades.library import Proxy
+
+# @notice This contract is only used to test the proxy functionnalities in tests.
```
Comment (hsouf) [inline tests/contracts/mock_aave_upgradeable_proxy.cairo:11]: Yes indeed, let's not forget to remove it!

**Review by hsouf — APPROVED** [review]

File: tests/unit_tests/protocol/libraries/helpers/test_bool_cmp.cairo [hunk 5]
```diff
@@ -45,3 +45,12 @@ func test_both{range_check_ptr}():
     assert false_3 = FALSE
     return ()
 end
+
+@external
+func test_not{range_check_ptr}():
+    let (true) = BoolCompare.not(0)
+    let (false) = BoolCompare.not(1)
```
Comment (EdgarBarrantes) [inline tests/unit_tests/protocol/libraries/helpers/test_bool_cmp.cairo:52]: Nitpick, and for me merge it without this, but you could've used `TRUE` and `FALSE` here.

**Review by EdgarBarrantes — APPROVED** [review]

**Pull request merged**